### PR TITLE
#done make Picasso runtime PRs more strict (because people click through in reviews instead of carefully reading)

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**
For runtime of Picasso, specifically lib.rs and governance.rs, we require not 2, but **3** reviewers **OR** require 2 but when one of reviewers from core team (assume most experienced in the substrate) to avoid adding security issues and not approved pallets to Picasso. So it just makes review one minimal step stricter for some files.

### CODE REVIEW

**Comment by dzmitry-lahoda** [comment]
@andyjsbell there is no task for it too - so it just change to make Picasso runtime more strict review to avoid adding security holes and data migrations issues.

**Review by KaiserKarel — APPROVED** [review]

File: .mergify.yml
Comment (vimukthi-git) [inline .mergify.yml:26]: What does this line do?

**Comment by andyjsbell** [comment]
@dzmitry-lahoda thanks for this.   I can probably infer from the code what will happen after this change but it could be a good idea to describe this in the PR description to allow us to review more efficiently.

File: .mergify.yml
Comment (dzmitry-lahoda) [reply .mergify.yml:26]: it avoids merge for wip hash tagged PRs:

https://en.wikipedia.org/wiki/Hashtag

Comment (dzmitry-lahoda) [reply .mergify.yml:26]: Actually it does same it did for other flow (10 lines above this #wip) , @vimukthi-git please look into whole small mergify yaml file. Imho it is hard to review if you do not look into what exists (there was already wip filter). Please skim over for DSL https://docs.mergify.com/conditions/ if something unclear.

**Comment by dzmitry-lahoda** [comment]
@andyjsbell 
`for runtime of Picasso, specifically lib.rs and governance.rs, require not 2, but 3 reviewers OR require 2 but when one of reviewers from core team (assume most experienced in the substrate)` to avoid adding security issues and not approved pallets to Picasso. So it just makes review one minimal step stricter for some files.

But, 
docs are here:
https://docs.mergify.com/conditions/
The assumption that DSL is good enough to be clear conditions (and there is

File: .mergify.yml
Comment (vimukthi-git) [inline .mergify.yml:20]: I would remove this as without a review from core or technical council, Picasso runtime must not be updated

Comment (vimukthi-git) [reply .mergify.yml:20]: But we can include a "or" condition to require review from technical council members

Comment (dzmitry-lahoda) [reply .mergify.yml:20]: You may update it in next PR. I did one step stricter. I do not do it as strict as possible. That is out of scope. So I do one step to improve. 

I am sure with my PR it is better then without. So it is not ideal. But still better. 

Settling ` technical council members` in git is out of scope of this PR and afaik not planned. So after it is done - feel free to change the rule.

Comment (dzmitry-lahoda) [reply .mergify.yml:20]: Suggesting using `core` only is good, but risky too.

Issue and required task with tech council members and core team mentioned in slack and meeting several times by me. So I would not do it.

Comment (vimukthi-git) [reply .mergify.yml:20]: >Suggesting using core only is good, but risky too.

The risk is worth taking for now IMO, as it has 3 of the members of the technical council now. I can add a condition later to require technical council review later.

Comment (dzmitry-lahoda) [reply .mergify.yml:20]: Imho I do not want to be person indulging work stoppage on people works by thinking that 3 tech member are really review this stuff (2 are likely not) - so only one and he is busy.   So it is risky. Also core has non overlapping meaning with tech collective. So it is not 1 = 1.

Comment (dzmitry-lahoda) [reply .mergify.yml:20]: Neither want to jump ahead and even badly assembled tech oracle. Final design is

1. each tech collective member must have identity (it can be nick name) and each such must have gpg key
2. any change to runtime triggers 
2.1. subkey dumps gpg keys form identity pallet
2.2. git checks gpg signatures on PR/branch/tag
2.3 if these do not apply - it fails

So whatever we can do to mergify will not replicate what is final achievable solution.

So current PR does improve things, but not blocks work. 

@vimukthi-git do you recall that we must gpg sign release? So actually we can avoid that action or reduce it security and embeeed into release pipeline same way we do with runtime changes.

But I do not want to spent time for issues. I just wanted to  do some more realistic rules and increase security in 30 minutes of work :) For any other improvements (with any changes risking blocking work done) it should go via CU.

Comment (dzmitry-lahoda) [reply .mergify.yml:20]: cc @andyjsbell just FYI . above some design for continuous improvement of our release and PR process (with automation). I did not digged GitLab runners of Parity, but at least per their release docs and tooling they are starting(ed) beginning of like this.

**Comment by andyjsbell** [comment]
> @andyjsbell `for runtime of Picasso, specifically lib.rs and governance.rs, require not 2, but 3 reviewers OR require 2 but when one of reviewers from core team (assume most experienced in the substrate)` to avoid adding security issues and not approved pallets to Picasso. So it just makes review one minimal step stricter for some files.
> 
> But, docs are here: https://docs.mergify.com/conditions/ The assumption that DSL is good enough to be clear conditions (and there is

Updated desc in PR, thanks!

File: .mergify.yml
Comment (andyjsbell) [reply .mergify.yml:20]: I think this is good @vimukthi-git, if you are happy then let's merge this one

**Review by andyjsbell — APPROVED** [review]
Thanks @dzmitry-lahoda

**Comment by dzmitry-lahoda** [comment]
> > @andyjsbell `for runtime of Picasso, specifically lib.rs and governance.rs, require not 2, but 3 reviewers OR require 2 but when one of reviewers from core team (assume most experienced in the substrate)` to avoid adding security issues and not approved pallets to Picasso. So it just makes review one minimal step stricter for some files.
> > But, docs are here: https://docs.mergify.com/conditions/ The assumption that DSL is good enough to be clear conditions (and there is
> 
> Updated desc in PR, thanks!

@andyjsbell just thought we all YAML programmers. YAML is marketed as human readable DSL for CI/CI/DevSecOps :)

**Comment by andyjsbell** [comment]
> > > @andyjsbell `for runtime of Picasso, specifically lib.rs and governance.rs, require not 2, but 3 reviewers OR require 2 but when one of reviewers from core team (assume most experienced in the substrate)` to avoid adding security issues and not approved pallets to Picasso. So it just makes review one minimal step stricter for some files.
> > > But, docs are here: https://docs.mergify.com/conditions/ The assumption that DSL is good enough to be clear conditions (and there is
> > 
> > 
> > Updated desc in PR, thanks!
> 
> @andyjsbell just thought we all YAML programmers. YAML is marketed as human readable DSL for CI/CI/DevSecOps :)

Of course @dzmitry-lahoda but I think it is still important to express the intention of what the change is in the desc of the PR.  We can still make mistakes in "human" like YAML and having a description allows us to validate assumptions 😄

File: .mergify.yml
Comment (vimukthi-git) [reply .mergify.yml:20]: @dzmitry-lahoda validation on the release by and signing it is not the time to find out that things are horribly broken. PR should allow catching any problems with the specific change as much as possible. I'm ok with the current change though.

**Review by vimukthi-git — APPROVED** [review]

File: .mergify.yml
Comment (vimukthi-git) [inline .mergify.yml:25]:
```suggestion
      - files ~= ^code\/parachain\/runtime\/picasso\/(lib\.rs|governance\.rs|xcmp\.rs)
```
Not sure why this file is skipped

**Comment by andyjsbell** [comment]
@dzmitry-lahoda can you hook this up to CU please?

**Comment by dzmitry-lahoda** [comment]
closing to align with other stuff in nix and src.

**Pull request closed without merging**
